src/lib/coda_base/ledger.ml

```diff
@@ -2,416 +2,510 @@ open Core
+module Db :
+  Merkle_ledger.Database_intf.S
+  with module Location = Location_at_depth
+  with module Addr = Location_at_depth.Addr
+  with type root_hash := Ledger_hash.t
+   and type hash := Ledger_hash.t
+   and type account := Account.t
+   and type key := Public_key.Compressed.t =
+  Database.Make (Key) (Account) (Hash) (Depth) (Location_at_depth) (Kvdb)
+    (Storage_locations)
+
+module Any_ledger :
+  Merkle_ledger.Any_ledger.S
+  with module Location = Location_at_depth
+  with type account := Account.t
+   and type key := Key.t
+   and type hash := Hash.t =
+  Merkle_ledger.Any_ledger.Make_base (Key) (Account) (Hash) (Location_at_depth)
+    (Depth)
+
+module Mask :
+  Merkle_mask.Masking_merkle_tree_intf.S
+  with module Location = Location_at_depth
+  with module Addr = Location_at_depth.Addr
+   and module Attached.Addr = Location_at_depth.Addr
+  with type account := Account.t
+   and type key := Key.t
+   and type hash := Hash.t
+   and type location := Location_at_depth.t
+   and type parent := Any_ledger.M.t =
+  Merkle_mask.Masking_merkle_tree.Make (Key) (Account) (Hash)
+    (Location_at_depth)
+    (Any_ledger.M)
+
+module Maskable :
+  Merkle_mask.Maskable_merkle_tree_intf.S
+  with module Location = Location_at_depth
+  with module Addr = Location_at_depth.Addr
+  with type account := Account.t
+   and type key := Key.t
+   and type hash := Hash.t
+   and type root_hash := Hash.t
+   and type unattached_mask := Mask.t
+   and type attached_mask := Mask.Attached.t
+   and type t := Any_ledger.M.t =
+  Merkle_mask.Maskable_merkle_tree.Make (Key) (Account) (Hash)
+    (Location_at_depth)
+    (Any_ledger.M)
+    (Mask)
+
+include Mask.Attached
+
+type maskable_ledger = t
+
+(* Mask.Attached.create () fails, can't create an attached mask directly
+   shadow create in order to create an attached mask
+*)
+let create ?directory_name () =
+  let maskable = Db.create ?directory_name () in
+  let casted = Any_ledger.cast (module Db) maskable in
+  let mask = Mask.create () in
+  Maskable.register_mask casted mask
+
+let with_ledger ~f =
+  let ledger = create () in
+  try
+    let result = f ledger in
+    destroy ledger ; result
+  with exn -> destroy ledger ; raise exn
+
+let packed t = Any_ledger.cast (module Mask.Attached) t
+
+let register_mask t mask = Maskable.register_mask (packed t) mask
+
+let unregister_mask_exn t mask = Maskable.unregister_mask_exn (packed t) mask
+
+(* TODO: Implement the serialization/deserialization *)
+let unattached_mask_of_serializable _ = failwith "unimplmented"
+
+let serializable_of_t _ = failwith "unimplented"
+
+type serializable = int [@@deriving bin_io]
+
+type unattached_mask = Mask.t
+
+type attached_mask = Mask.Attached.t
+
+(* inside MaskedLedger, the functor argument has assigned to location, account, and path
+   but the module signature for the functor result wants them, so we declare them here *)
+type location = Location.t
+
+let fold_until t ~init ~f ~finish =
```

Comment:
can we make an issue for this to not allocate a list? I feel like the expectation with a fold is that e.g. `fold t ~f:(fun acc _ -> acc + 1) ~init:0` should use a constant amount of memory